### PR TITLE
[FLINK-19134][python] Add BasicArrayTypeInfo and coder for PrimitiveArrayTypeInfo for Python DataStream API.

### PR DESCRIPTION
**Pull request by shuiqiangchen**
## What is the purpose of the change

Add BasicArrayTypeInfo and coder for PrimitiveArrayTypeInfo.

## Brief change log

- Added BasicArrayTypeInfo for arrays boxed primitive types.
- Added coder for PrimitiveArrayTypes.

## Verifying this change

This pull request has test case covered by test_primitive_array_type_info() and test_basic_array_type_info() in test_data_stream.py.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? ( not documented)

### CODE REVIEW

File: flink-python/pyflink/common/typeinfo.py [hunk 1]
```diff
@@ -209,6 +209,60 @@ def CHAR_PRIMITIVE_ARRAY_TYPE_INFO():
             .PrimitiveArrayTypeInfo.CHAR_PRIMITIVE_ARRAY_TYPE_INFO)
 
 
+class BasicArrayTypeInfo(TypeInformation, ABC):
+    """
+    A TypeInformation for arrays boxed primitive types (Integer, Long, Double, ...).
```
Comment (HuangXingBo) [inline flink-python/pyflink/common/typeinfo.py:214]:
```suggestion
    A TypeInformation for arrays of boxed types (Integer, Long, Double, ...).
```

File: flink-python/pyflink/common/typeinfo.py [hunk 2]
```diff
@@ -384,6 +438,27 @@ def PRIMITIVE_ARRAY(element_type: TypeInformation):
         else:
             raise TypeError("Invalid element type for a primitive array.")
 
+    @staticmethod
+    def BASIC_ARRAY(element_type: TypeInformation):
```
Comment (HuangXingBo) [inline flink-python/pyflink/common/typeinfo.py:442]: Adds the annotation

File: flink-python/pyflink/common/typeinfo.py [hunk 3]
```diff
@@ -384,6 +438,27 @@ def PRIMITIVE_ARRAY(element_type: TypeInformation):
...
+        elif element_type == Types.LONG():
+            return BasicArrayTypeInfo.LONG_ARRAY_TYPE_INFO()
+        elif element_type == Types.FLOAT():
+            return BasicArrayTypeInfo.FLOAT_ARRAY_TYPE_INFO()
+        elif element_type == Types.DOUBLE():
+            return BasicArrayTypeInfo.DOUBLE_ARRAY_TYPE_INFO()
+        elif element_type == Types.CHAR():
+            return BasicArrayTypeInfo.CHAR_ARRAY_TYPE_INFO()
+        else:
+            raise TypeError("Invalid element type for a primitive array.")
```
Comment (HuangXingBo) [inline flink-python/pyflink/common/typeinfo.py:460]:
```suggestion
            raise TypeError("Invalid element type for a boxed array.")
```

File: flink-python/pyflink/common/typeinfo.py [hunk 4]
```diff
@@ -384,6 +438,27 @@ def PRIMITIVE_ARRAY(element_type: TypeInformation):
...
+        elif element_type == Types.INT():
+            return BasicArrayTypeInfo.INT_ARRAY_TYPE_INFO()
+        elif element_type == Types.LONG():
+            return BasicArrayTypeInfo.LONG_ARRAY_TYPE_INFO()
+        elif element_type == Types.FLOAT():
+            return BasicArrayTypeInfo.FLOAT_ARRAY_TYPE_INFO()
+        elif element_type == Types.DOUBLE():
+            return BasicArrayTypeInfo.DOUBLE_ARRAY_TYPE_INFO()
+        elif element_type == Types.CHAR():
+            return BasicArrayTypeInfo.CHAR_ARRAY_TYPE_INFO()
```
Comment (HuangXingBo) [inline flink-python/pyflink/common/typeinfo.py:458]: Do you forget to add BasicArrayTypeInfo.STRING_ARRAY_TYPE_INFO ?

File: flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py [hunk 5]
```diff
@@ -187,6 +187,24 @@ def __repr__(self):
         return 'ArrayCoderImpl[%s]' % repr(self._elem_coder)
 
 
+class PrimitiveArrayCoderImpl(StreamCoderImpl):
```
Comment (HuangXingBo) [inline flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py:190]: I think we need to add related unit test of coders in test_coders.py and test_fast_coders.py

**Review by HuangXingBo — COMMENTED** [review]
@shuiqiangchen Thanks a lot for the PR. Looks good overall. I have left some comments.

**Comment by shuiqiangchen** [comment]
@HuangXingBo Thanks a lot for your comments, I have updated the PR according to your suggestions, please have a look at it.

File: flink-python/src/main/java/org/apache/flink/datastream/runtime/typeutils/python/PythonTypeUtils.java [hunk 6]
```diff
@@ -308,6 +366,14 @@ public static TypeSerializer typeInfoSerializerConverter(TypeInformation typeInf
 						.map(f -> typeInfoSerializerConverter(f)).toArray(TypeSerializer[]::new);
 					return new TupleSerializer(Tuple.getTupleClass(tupleTypeInfo.getArity()), fieldTypeSerialzers);
 				}
+
+				if (typeInformation instanceof BasicArrayTypeInfo){
```
Comment (HuangXingBo) [inline flink-python/src/main/java/org/apache/flink/datastream/runtime/typeutils/python/PythonTypeUtils.java:370]: Do we need to add the corresponding test content in `testTypeInfotoSerializerConverter`

File: flink-python/pyflink/fn_execution/coders.py [hunk 7]
```diff
@@ -226,13 +226,24 @@ class ArrayCoder(CollectionCoder):
...
 
     def __init__(self, elem_coder):
-        self._elem_coder = elem_coder
         super(ArrayCoder, self).__init__(elem_coder)
 
     def get_impl(self):
         return coder_impl.ArrayCoderImpl(self._elem_coder.get_impl())
 
 
+class PrimitiveArrayCoder(CollectionCoder):
```
Comment (HuangXingBo) [inline flink-python/pyflink/fn_execution/coders.py:235]: Now that we have introduced `PrimitiveArrayCoder`, can we rename `ArrayCoder` to `BasicArrayCoder`

File: flink-python/pyflink/datastream/tests/test_data_stream.py [hunk 8]
```diff
@@ -557,6 +557,44 @@ def assert_chainable(j_stream_graph, expected_upstream_chainable,
...
+                                      (3, [3.1, 3.2, None], ['hello', 'hi', None])],
+                                      type_info=Types.ROW([Types.INT(),
+                                                           Types.BASIC_ARRAY(Types.FLOAT()),
+                                                           Types.BASIC_ARRAY(Types.STRING())]))
+
+        ds.map(lambda x: x, output_type=Types.ROW([Types.INT(),
+                                                   Types.BASIC_ARRAY(Types.FLOAT()),
+                                                   Types.BASIC_ARRAY(Types.STRING())]))\
+            .add_sink(self.test_sink)
+        self.env.execute("test array type info")
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/tests/test_data_stream.py:588]:
```suggestion
        self.env.execute("test basic array type info")
```

File: flink-python/pyflink/datastream/tests/test_data_stream.py [hunk 9]
```diff
@@ -557,6 +557,44 @@ def assert_chainable(j_stream_graph, expected_upstream_chainable,
...
+    def test_primitive_array_type_info(self):
+        ds = self.env.from_collection([(1, [1.1, 1.2, 1.30]), (2, [2.1, 2.2, 2.3]),
+                                      (3, [3.1, 3.2, 3.3])],
+                                      type_info=Types.ROW([Types.INT(),
+                                                           Types.PRIMITIVE_ARRAY(Types.FLOAT())]))
+
+        ds.map(lambda x: x, output_type=Types.ROW([Types.INT(),
+                                                   Types.PRIMITIVE_ARRAY(Types.FLOAT())]))\
+            .add_sink(self.test_sink)
+        self.env.execute("test array type info")
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/tests/test_data_stream.py:569]:
```suggestion
        self.env.execute("test primitive array type info")
```

File: flink-python/pyflink/datastream/tests/test_data_stream.py [hunk 10]
```diff
@@ -557,6 +557,44 @@ def assert_chainable(j_stream_graph, expected_upstream_chainable,
...
+                                                   Types.BASIC_ARRAY(Types.STRING())]))\
+            .add_sink(self.test_sink)
+        self.env.execute("test array type info")
+        results = self.test_sink.get_results()
+        expected = ['1,[1.1, null, 1.3],[null, hi, flink]',
+                    '2,[null, 2.2, 2.3],[hello, null, flink]',
+                    '3,[3.1, 3.2, null],[hello, hi, null]']
+        results.sort()
+        expected.sort()
+        print(results)
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/tests/test_data_stream.py:595]: remove debugging information

**Review by HuangXingBo — COMMENTED** [review]
@shuiqiangchen Thanks a lot for the update. I have left another some comments.

**Comment by HuangXingBo** [comment]
@shuiqiangchen Thanks a lot for the update. The PR already looks quite good to me.
Let's also wait for @dianfu to take a look.

**Comment by shuiqiangchen** [comment]
@flinkbot run azure

File: flink-python/pyflink/common/typeinfo.py [hunk 11]
```diff
@@ -384,6 +444,35 @@ def PRIMITIVE_ARRAY(element_type: TypeInformation):
...
+        elif element_type == Types.STRING():
+            return BasicArrayTypeInfo.STRING_ARRAY_TYPE_INFO()
+        else:
+            raise TypeError("Invalid element type for a boxed primitive array.")
```
Comment (dianfu) [inline flink-python/pyflink/common/typeinfo.py:474]:
```suggestion
            raise TypeError("Invalid element type for a boxed primitive array: %s" % str(element_type))
```

File: flink-python/pyflink/fn_execution/beam/beam_coders.py [hunk 12]
```diff
@@ -194,7 +194,7 @@ def _to_data_type(field_type):
                                                field_type.nullable)
             elif field_type.type_name == flink_fn_execution_pb2.Schema.TIMESTAMP:
                 return TimestampType(field_type.timestamp_info.precision, field_type.nullable)
-            elif field_type.type_name == flink_fn_execution_pb2.Schema.ARRAY:
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/beam/beam_coders.py:197]: should we also add PRIMITIVE_ARRAY?

File: flink-python/pyflink/fn_execution/coders.py [hunk 13]
```diff
@@ -32,7 +32,7 @@
 __all__ = ['RowCoder', 'BigIntCoder', 'TinyIntCoder', 'BooleanCoder',
            'SmallIntCoder', 'IntCoder', 'FloatCoder', 'DoubleCoder',
            'BinaryCoder', 'CharCoder', 'DateCoder', 'TimeCoder',
-           'TimestampCoder', 'ArrayCoder', 'MapCoder', 'DecimalCoder']
+           'TimestampCoder', 'BasicArrayCoder', 'MapCoder', 'DecimalCoder']
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coders.py:35]: Also expose PrimitiveArrayCoder?

File: flink-python/pyflink/fn_execution/coders.py [hunk 14]
```diff
@@ -467,8 +478,8 @@ def from_proto(field_type):
     if field_type_name == type_name.LOCAL_ZONED_TIMESTAMP:
         timezone = pytz.timezone(os.environ['table.exec.timezone'])
         return LocalZonedTimestampCoder(field_type.local_zoned_timestamp_info.precision, timezone)
-    elif field_type_name == type_name.ARRAY:
-        return ArrayCoder(from_proto(field_type.collection_element_type))
+    elif field_type_name == type_name.BASIC_ARRAY:
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coders.py:481]: Should we also support PRIMITIVE_ARRAY here?

**Review by dianfu — COMMENTED** [review]
@shuiqiangchen Thanks a lot for the PR. Only a few minor comments.

File: flink-python/pyflink/fn_execution/coders.py (hunk 14 above)
Comment (shuiqiangchen) [reply flink-python/pyflink/fn_execution/coders.py:481]: In SQL/ Table API, the array type is the BasicArrayType,  PrimitiveArrayType is not considered.

File: flink-python/pyflink/fn_execution/beam/beam_coders.py (hunk 12 above)
Comment (shuiqiangchen) [reply flink-python/pyflink/fn_execution/beam/beam_coders.py:197]: In SQL/ Table API, the array type is the BasicArrayType, PrimitiveArrayType is not considered.

**Review by dianfu — APPROVED** [review]

**Comment by shuiqiangchen** [comment]
@flinkbot run azure

**Comment by dianfu** [comment]
The test failure is not related to this PR. Merging...

**Pull request closed without merging**